### PR TITLE
Support for Qute Type-safe Message Bundles

### LINKED ISSUE #800: Initialize support for Qute `Type-safe Message Bundles`
See https://quarkus.io/guides/qute-reference#type-safe-message-bundles

### PR DESCRIPTION
**Pull request by angelozerr**
Support for Qute Type-safe Message Bundles

Fixes #800

### CODE REVIEW

**Comment by angelozerr** [comment]
The support for Qute Type-safe Message Bundles starts working with Java interfaces (but not with properties file).

See the following demo:

MessageWithJavaInterfaceDemo

@mkouba @FroMage I have started the support for Qute Type-safe Message Bundles which works only with Java interface. I need to support properties file.

Any feedback are welcome if you need some feature.

**Comment by angelozerr** [comment]
@mkouba I have tried to play with @MessageBundle and I have several questions:

 * could we define 2 interfaces wich implements @MessageBundle  and the other @MessageBundle("msg1"). I tried it and to write expression like {msg:hello} and {msg1:hello} but it doesnt work. 
 * according my test, you need to write an interface which implement @MessageBundle even if you wish to use msg:message (with a key). Is it normal?

**Comment by angelozerr** [comment]
I have improved the support and now you can see message content as inlay hint in the template:

image

**Comment by angelozerr** [comment]
> I think that we should display the message template and not the evaluaded message. Mainly because we are not always able to evaluate the message template and then it might become confusing... (the {msg:hello_name(person.name)}  [Hello person.name!] example).

Ok thanks for your feedback @mkouba !

**Comment by angelozerr** [comment]
I need to clean my code and write tests. This PR provides the first support for Qute messages but it need to be improved with:

 * support as inlay hint message when it is defined in a msg.properties file.
 * code action to create an AppMessages when it doesn't exists.
 * support msg:message completion
 * other?

I will need that in separate PRs to avoid having a big PR.

@datho7561 @JessicaJHee you can start to play with this PR. Once tests and code will be cleaned I will notice you.

**Comment by angelozerr** [comment]
To play with Qute messages, I suggest that you use the qute-messages that I will use for Qute JDT tests.

**Comment by angelozerr** [comment]
Please note that I have encountered 2 bugs in Qute messages support:

 * https://github.com/quarkusio/quarkus/issues/32354
 * https://github.com/quarkusio/quarkus/issues/32355

**Comment by angelozerr** [comment]
> I don't know. I was going to answer the other way, and display the text substituted. At a point (in the view) where we don't know what {name} maps to, because we don't know the name of the parameters, I find it more useful to display Hello Lucie for constants, and Hello {person.name} for non-constant expressions.

It was my original idea by @mkouba think we should not evaluate the message https://github.com/redhat-developer/quarkus-ls/pull/845#issuecomment-1493927666

**Comment by angelozerr** [comment]
@datho7561 @JessicaJHee the PR can be reviewed now, code is clean and tests exists now.

**Review by JessicaJHee — COMMENTED** [review]
I played around and looks great! I just noticed some very minor things.

I think this feature you mentioned would be very nice to have in a future PR:

> * support as inlay hint message when it is defined in a msg.properties file.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TypeSafeMessageBundlesSupport.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TypeSafeMessageBundlesSupport.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/resolvers/MessageValueResolver.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/resolvers/MessageValueResolver.java]: fixed

**Comment by angelozerr** [comment]
> I played around and looks great! 

Thanks!

> I think this feature you mentioned would be very nice to have in a future PR:

It is in my plan, but this PR is already big, so I prefer providing message support in separate PRs to have easier review.

**Comment by angelozerr** [comment]
@JessicaJHee I have improved inlay hint and now you click on it to edit the Java message by going to the java method. 

MessageInlayHintClickable

It requires to improve it to set the location to the @Message annotation content and not the Java method but we can do that in a separate PR (if you are interested you could work on it).

Once message properties support will be available https://github.com/redhat-developer/quarkus-ls/issues/847 my idea is to provide the capability to open the msg.properties file.

**Comment by angelozerr** [comment]
@datho7561 @JessicaJHee can we merge this PR?

**Comment by datho7561** [comment]
I haven't had a chance to look at it yet, but I know that Jessica reviewed it. Do you think it's ready to merge, Jessica?

**Pull request merged**